### ISSUE TITLE
Support for Windows Projected File System (ProjFS)

### ISSUE THREAD

**Issue by jeanga** [ufrisk/MemProcFS#28]
Hi!

Did you ever consider using [Windows Projected File System (ProjFS)](https://docs.microsoft.com/en-us/windows/win32/projfs/projected-file-system) for MemProcFS?

Thanks!

**Comment by ufrisk** [comment]
I remember I looked at it when it was brand new, but I decided against using it at that time and then I forgot about it.

I think my reasoning back then was that it did not provide any real benefit over Dokan which I use now; and back then it was only available in the most recent versions of Windows - meaning if I would have been migrating to it I would have been locking out large amounts of the users.

That is no longer the case though; which makes it more attractive. I also believe it's probably not going to be that hard to implement it. It's built into Windows, which is nice; but there is still some work for me to add support for it. But what would be the benefit over continuing to use Dokan as I do now?

**Comment by jeanga** [comment]
Well, Dokan, if I remember correctly, involves the install of their custom driver.
As security professionals, I think a lot of your users would love to avoid that step.

To me, not relying on Dokan's driver security to stay safe would be a huge plus for me to install and use MemProcFS.

**Comment by ufrisk** [comment]
Yes, it does, and it's fairly widely used so I'm not too worried. If this is a huge concern running in a VM is an option.

But I'll look into it; it's not a bad idea; doing dual support for Dokan/ProjFS in the same binary with auto-detect shouldn't be too hard I hope; but it will take some time to implement for sure. I'll look into it and put it up as an enhancement request for the future (like in later this year, but not as far in the future as next year).

**Comment by ufrisk** [comment]
I have finally looked into this. Apologies for the huve amount of time passed; it just haven't been on the top of my prio list. I even was in the early stages of creating a PoC for ProjFs when I noticed there is no API to intercept writes to a file.

MemProcFS relies on the ability to write to files for quite a few features. As ProjFs stands now it's unfortunately not suitable as a replacement to Dokan.

More info here: https://github.com/microsoft/ProjFS-Managed-API/issues/30

Due to these limitations in ProjFs I'm closing this issue.

it would have been nice to have built-in Windows support for the virtual file system.

But Dokan is a widely used and stable project so I wouldn't be worried about it; still it's a shame ProjFs just isn't suitable...
